cli/azd/internal/cmd/add/add.go

```diff
@@ -0,0 +1,379 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package add
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"log"
+	"os"
+	"path/filepath"
+	"slices"
+	"strings"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore/arm"
+	"github.com/azure/azure-dev/cli/azd/cmd/actions"
+	"github.com/azure/azure-dev/cli/azd/internal/repository"
+	"github.com/azure/azure-dev/cli/azd/pkg/account"
+	"github.com/azure/azure-dev/cli/azd/pkg/alpha"
+	"github.com/azure/azure-dev/cli/azd/pkg/environment"
+	"github.com/azure/azure-dev/cli/azd/pkg/environment/azdcontext"
+	"github.com/azure/azure-dev/cli/azd/pkg/infra"
+	"github.com/azure/azure-dev/cli/azd/pkg/input"
+	"github.com/azure/azure-dev/cli/azd/pkg/osutil"
+	"github.com/azure/azure-dev/cli/azd/pkg/output"
+	"github.com/azure/azure-dev/cli/azd/pkg/output/ux"
+	"github.com/azure/azure-dev/cli/azd/pkg/project"
+	"github.com/azure/azure-dev/cli/azd/pkg/prompt"
+	"github.com/azure/azure-dev/cli/azd/pkg/workflow"
+	"github.com/azure/azure-dev/cli/azd/pkg/yamlnode"
+	"github.com/braydonk/yaml"
+	"github.com/fatih/color"
+	"github.com/spf13/cobra"
+)
+
+func NewAddCmd() *cobra.Command {
+	return &cobra.Command{
+		Use:   "add",
+		Short: fmt.Sprintf("Add a component to your application. %s", output.WithWarningFormat("(Alpha)")),
+	}
+}
+
+type AddAction struct {
+	azd              workflow.AzdCommandRunner
+	azdCtx           *azdcontext.AzdContext
+	env              *environment.Environment
+	envManager       environment.Manager
+	subManager       *account.SubscriptionsManager
+	alphaManager     *alpha.FeatureManager
+	creds            account.SubscriptionCredentialProvider
+	rm               infra.ResourceManager
+	appInit          *repository.Initializer
+	armClientOptions *arm.ClientOptions
+	prompter         prompt.Prompter
+	console          input.Console
+}
+
+var composeFeature = alpha.MustFeatureKey("compose")
+
+func (a *AddAction) Run(ctx context.Context) (*actions.ActionResult, error) {
+	if !a.alphaManager.IsEnabled(composeFeature) {
+		return nil, fmt.Errorf(
+			"compose is currently under alpha support and must be explicitly enabled."+
+				" Run `%s` to enable this feature", alpha.GetEnableCommand(composeFeature),
+		)
+	}
+
+	prjConfig, err := project.Load(ctx, a.azdCtx.ProjectPath())
+	if err != nil {
+		return nil, err
+	}
+
+	selectMenu := a.selectMenu()
+	slices.SortFunc(selectMenu, func(a, b Menu) int {
+		return strings.Compare(a.Label, b.Label)
+	})
```

Comment:
nit.  Should `selectMenu()` return a sorted-by-label slice?

Reply:
I do prefer it this way -- I'm setting up `Menu` as a data contract of items to display, but I'd prefer not making it a UI component for now. The UI component is actually within `add.go` that can decide how to render elements based on what is best for the presentation.